akka-actor/src/main/scala-2.13+/akka/util/ByteString.scala

```diff
@@ -754,10 +781,21 @@ sealed abstract class ByteString
    */
   protected[ByteString] def toArray: Array[Byte] = toArray[Byte]
 
-  override def toArray[B >: Byte](implicit arg0: ClassTag[B]): Array[B] = iterator.toArray
+  final override def toArray[B >: Byte](implicit arg0: ClassTag[B]): Array[B] = {
+    // super uses byteiterator
+    val array = new Array[B](size)
+    copyToArray(array, 0, size)
+    array
+  }
+
+  final override def copyToArray[B >: Byte](xs: Array[B], start: Int): Int = {
+    // super uses byteiterator
+    copyToArray(xs, start, size.min(xs.size))
+  }
 
+  // optimized in all subclasses, avoiding usage of the iterator to save allocations/transformations
   override def copyToArray[B >: Byte](xs: Array[B], start: Int, len: Int): Int =
-    iterator.copyToArray(xs, start, len)
+    throw new UnsupportedOperationException("Method copyToArray is not implemented in ByteString")
```

Comment:
:+1:

Reply:
is that because it should alway be implemented by subclasses of ByteString?

Reply:
Yes, if we missed it we'd get the crappy iterator byte-for-byte one